pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,292 @@
+
+### Write Invalidate - read path
+Read requests are handled by the generic read function, as in
+[**Write Through read path**](/ocf_io_engines.html#write-through---read-path).
+
+
+## Pass Through (PT)
+This engine is used to bypass the cache device. However, in some circumstances, 
+a complete bypass is not possible because dirty data may exist in the cache.
+
+Used by the [**Pass Through cache mode**](/cache_configuration#pass-through)
+as well as a fallback engine for corner cases.
+
+These are the functions used:
+~~~~
+[OCF_IO_PT_IF] = {
+        .read = ocf_read_pt,
+        .write = ocf_write_wi,
+        .name = "Pass Through",
+},
+~~~~
+
+### Pass Through - write path
+Write requests are handled by the [**Write Invalidate write path**](/ocf_io_engines.html#write-invalidate---write-path).
+
+### Pass Through - read path
+In most cases, the data is read from the core device. As the cache is bypassed, 
+backfill is not performed - reads are not accelerated in this case.
+
+However, if the data is hit and clean in the cache, the data read is read from
+the cache device. If the data is dirty, a cleaner is invoked and the request is 
+reissued. 
+
+If a sequential cutoff threshold is hit and the data is 100% dirty, a regular 
+[**Write Through read**](/ocf_io_engines.html#write-through---read-path)
+is performed.
+
+
+## Write Only (WO)
+This engine used in a read path in write-only cache mode reads the data
+from a suitable device (either cache or core) but does not change the data 
+state in the cache. 
+
+Write Only cache mode is used as an accelerated write buffer. It is meant to be 
+used with partitioning based on write life-time hints for optimal performance.
+See [**write only cache mode configuration**](/cache_configuration.html#write-only)  
+These are the function used:
+~~~~
+[OCF_IO_WO_IF] = {
+        .read = ocf_read_wo,
+        .write = ocf_write_wb,
+        .name = "Write Only",
+},
+~~~~
+
+### Write Only - write path
+Write requests are handled by the [**Write Back write path**](/ocf_io_engines.html#write-back---write-path).
+
+### Write Only - read path
+Reads do not change the cache state and backfill is not performed.
+
+Because of that, some of the requested data may be read from the core, but a 
+valid, clean data may be read from the cache device.
+
+
+## Fast
+This engine was implemented for historical reasons - BIO vs request-based 
+IO interfaces in Linux kernel spawned a change in the IO path implementation.
+
+In contrast to a regular CAS IO handling by using request queues and I/O threads,
+Fast engine, or Fast Path, executes engine logic synchronously.
+
+I/Os handled by Fast Path are 
+* read and write hit in WB and WO modes
+* other modes except PT - read hit only
+
+
+## Direct To Core (D2C)
+Direct To Core, or D2C, was implemented for handling a detached cache drive 
+state of CAS. In this mode, all read and write I/O requests go directly to the
+core device.
```

Comment:
Cacheline status metadata is not referenced or modified, as it is expected to be deallocated in detached cache state.